### DISCUSSION TITLE
No validation loss ?

### DISCUSSION THREAD

**Discussion by ssbagalkar** [STVIR/pysot#443] [General] [answered]
I have been trying to reproduce some of the results as well as trying to train on a custom dataset. While, the training runs without any issues, there is no validation loss reported, either on command line or on tensorboard. A little digging helped me to understand that inside `pysot/models/model_builder.py` only the `total_loss`, `cls_loss` and `loc_loss` is reported. Can we add validation losses to this code? If so, how and can anyone give me any tips. I would love to contribute as well

    def forward(self, data):
        """ only used in training
        """
        template = data['template'].cuda()
        search = data['search'].cuda()
        label_cls = data['label_cls'].cuda()
        label_loc = data['label_loc'].cuda()
        label_loc_weight = data['label_loc_weight'].cuda()

        # get feature
        zf = self.backbone(template)
        xf = self.backbone(search)
        if cfg.MASK.MASK:
            zf = zf[-1]
            self.xf_refine = xf[:-1]
            xf = xf[-1]
        if cfg.ADJUST.ADJUST:
            zf = self.neck(zf)
            xf = self.neck(xf)
        cls, loc = self.rpn_head(zf, xf)

        # get loss
        cls = self.log_softmax(cls)
        cls_loss = select_cross_entropy_loss(cls, label_cls)
        loc_loss = weight_l1_loss(loc, label_loc, label_loc_weight)

        outputs = {}
        outputs['total_loss'] = cfg.TRAIN.CLS_WEIGHT * cls_loss + \
            cfg.TRAIN.LOC_WEIGHT * loc_loss
        outputs['cls_loss'] = cls_loss
        outputs['loc_loss'] = loc_loss

        if cfg.MASK.MASK:
            # TODO
            mask, self.mask_corr_feature = self.mask_head(zf, xf)
            mask_loss = None
            outputs['total_loss'] += cfg.TRAIN.MASK_WEIGHT * mask_loss
            outputs['mask_loss'] = mask_loss
        return outputs

**Comment by ZhiyuanChen** [comment, marked as answer]
It is the forward function, so if you call it in training, you get training loss, and if you call it in validation, you get validation loss
